### ISSUE TITLE
[TSX] Mixup between TS Generics and JSX syntax

### ISSUE THREAD

**Issue by ephemer** [sublimehq/Packages#2529]
- Sublime Version: 4086
- OS Version: Mac OS 11.0 (beta 6)

But using `Packages` at latest master (commit b375218f)

May be related to #2515 but I'm not sure whether it's exactly the same issue.

The following code is a minimally-reproducible version of the issue:

``` ts
import React from "react";
const Component: React.FC<any> = props => null;
// anything here will not be highlighted at all
```

Removing the generic type parameter on FC fixes highlighting (but breaks types). Substituting `React.FC<...>` with a variable does _not_ fix the issue however.

``` ts
type Sub = React.FC<any>;
const Component: Sub = props => null;
// still broken here
```

I assume it's because `<any>` is considered to be a JSX element rather than a type parameter.

Real-world example: 

<img width="619" alt="image" src="https://user-images.githubusercontent.com/5485935/93895211-de965980-fcef-11ea-93d4-580737cba9e0.png">

**Comment by Thom1729** [comment]
Thanks for the report. The linked PR should fix it. The root problem was that in a dotted type like `React.FC`, the syntax was in the wrong state to handle generics (also, the `[]` modifier or a bracket property access).

**Comment by ivanddm** [comment]
I'm facing this issue right now (ST4 4112 MacOS). Should I open a new issue?

Basically, `React.FC<...>` breaks the syntax highlighting as well as the commenting command, while `React.FC` without `<>` does not:

![image](https://user-images.githubusercontent.com/20104703/127907165-9050367d-a692-40a5-849e-b31897734ccb.png)

instead of: 

![image](https://user-images.githubusercontent.com/20104703/127907255-d38ae435-0109-4ccb-bef2-20d5467f78d4.png)

**Comment by ivanddm** [comment]
Well, I realized I was still using JSCustoms (from ST3). So I enabled the builtin syntax for TSX and it working now. Sorry about that.

**Comment by Thom1729** [comment]
FYI, it should work fine on the current version of JS Custom as well. If it doesn't, I'd appreciate if you would open a bug report.

**Comment by ivanddm** [comment]
Thanks for letting me know. I just checked setting JSCustoms in tsx files. I'm getting the following with the default syntaxes that are installed when you install JSCustom:

- JS Custom Default: working for syntax of regular JS-JSX-TS-TSX, not for commenting the JSX part (adds // intead of {/* */}).
- JS Custom React: working for JS-JSX-TS-TSX but uncommenting does not remove the curly braces.
- JS Custom Typescript: working for syntax of regular JS-JSXTS-TSX, not for commenting the JSX part (adds // intead of {/* */})
- JS Custom Typescript (JSX): not working at all, same issue as above.

I was going to ask you which one should I prefer, whether built-in JS-JSX-TS-TSX syntax that comes with ST4 or JSCustoms, however, based on what I can see, the new built-in syntaxes do proper commenting and uncommenting of JSX (they also remove the curly braces when uncommenting).

I will open a bug report in JS Custom based on this. Thanks in advance for your support

**Comment by ivanddm** [comment]
**Update**: Actually, although I had reinstalled JSCustoms, the old syntaxes were still in my package folder 🤦🏻‍♂️. After rebuilding JS Custom syntaxes, I can say that they work as expected just like the ST4 built-in ones.

Thanks again!

**Comment by mzaharie** [comment]
I don't have JSCustoms installed and the generics on arrow function still break the syntax highligher. Look:
<img width="603" alt="Screenshot 2021-08-18 at 11 36 31" src="https://user-images.githubusercontent.com/5528724/130646148-72799c26-7ee9-4fc7-a6e0-3879594bfa31.png">

**Comment by Thom1729** [comment]
@mzaharie This is due to #2660, which was closed just yesterday by #2923. The fix should presumably be in the next dev build, or you can check out the latest version manually if you like. I'm also going to work today on getting the fix (and other recent JavaScript family improvements) into JS Custom.
